internal/cfgmgr/cfgmgr.go

```diff
@@ -0,0 +1,163 @@
+		var devNodeInst uint32
+		err = cmLocateDevNode(&devNodeInst, id, _CM_LOCATE_DEVNODE_NORMAL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to locate device node for %s", id)
+		}
+		propertyType := uint32(0)
+		propertyBufferSize := uint32(0)
+
+		// get the size of the property buffer by querying with a nil buffer and zeroed propertyBufferSize
+		err = cmGetDevNodeProperty(devNodeInst, devPKeyDeviceLocationPaths, &propertyType, nil, &propertyBufferSize, 0)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get property buffer size of devnode query for %s with", id)
+		}
+
+		// get the property with the resulting propertyBufferSize
+		tempPropertyBufSize := propertyBufferSize
+		propertyBuffer := make([]uint16, propertyBufferSize*2)
+		err = cmGetDevNodeProperty(devNodeInst, devPKeyDeviceLocationPaths, &propertyType, &propertyBuffer[0], &propertyBufferSize, 0)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get location path property from device node for %s with", id)
+		}
+		if propertyType != _DEVPROP_TYPE_STRING_LIST {
+			return nil, fmt.Errorf("expected to return property type DEVPROP_TYPE_STRING_LIST %d, instead got %d", _DEVPROP_TYPE_STRING_LIST, propertyType)
+		}
+		if propertyBufferSize > tempPropertyBufSize {
+			return nil, fmt.Errorf("location path is too large for the buffer, size %d", propertyBufferSize)
+		}
+		formattedResult, err := convertFirstNullTerminatedValueToString(propertyBuffer[:propertyBufferSize])
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, formattedResult)
+	}
+
+	return result, nil
+}
+
+// helper function that finds the first \u0000 rune and returns the wide string as a regular go string
+func convertFirstNullTerminatedValueToString(buf []uint16) (string, error) {
+	r := utf16.Decode(buf)
+	converted := string(r)
+	zerosIndex := strings.IndexRune(converted, '\u0000')
+	if zerosIndex != -1 {
+		return "", fmt.Errorf("cannot convert value to string, malformed data passed: %v", buf)
+	}
+	return converted, nil
+}
+
+func GetChildrenFromInstanceIDs(parentIDs []string) ([]string, error) {
+	var result []string
+	for _, id := range parentIDs {
+		pszFilterParentID := []byte(id)
+		children, err := getDeviceIDList(&pszFilterParentID[0], _CM_GETIDLIST_FILTER_BUSRELATIONS)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, children...)
+	}
+	return result, nil
+}
+
+func getDeviceIDList(pszFilter *byte, ulFlags uint32) ([]string, error) {
+	listLength := uint32(0)
+	if err := cmGetDeviceIDListSize(&listLength, pszFilter, ulFlags); err != nil {
+		return nil, err
+	}
+	if listLength == 0 {
+		return []string{}, nil
+	}
+	buf := make([]byte, uint32(listLength))
+	if err := cmGetDeviceIDList(pszFilter, &buf[0], uint32(listLength), ulFlags); err != nil {
+		return nil, err
+	}
+
+	return convertPZZSTRToSlice(buf), nil
+}
+
+// convertPZZSTRToSlice is a helper function used to convert the contents of a
+// PZZSTR to a string slice
```

Comment:
What is a PZZSTR?

Reply:
That's a fair point. It's a C++ data type that contains a set of null terminated strings with an extra null at the end. I'll change this to something more descriptive since it's difficult to find what that is.